File: Project2/julia/jacobi.jl
module Jacobi
export offdiagmax, jacobirotate!, jacobi!

#=
This module provides different methods to find the eigenvalues for
matrices. These are taken verbatim from
https://compphysics.github.io/ComputationalPhysics/doc/pub/eigvalues/pdf/eigvalues-print.pdf
and only adapted to Julia.
=#

function jacobi!(A::Array{Float64}; ɛ=1e-10, maxiter=1e8)
    # Sets up the algorithm and begins a loop finding the max non diagonal element
    # before performing Given's rotations. It terminates if all non diagonal elements
    # are less than ε or if maxiter number of iterations are reached.
    # @Input: A, the array to find the eigenvalues of. To save memory, the
    #         array is overwritten with the eigenvalues along the diagonal.
    #         ε, tolerance of non diagonal elements
    #         maxiter, maximum number of iterations before breaking the loop
    # @Return: The eigenvectors and the number of iterations that was needed
    #          to make all off diagonals less than ε
    R = eye(A)
    iterations = 0
    maxnondiag = offdiagmax(A)[1]
    while (maxnondiag > ɛ && iterations ≤ maxiter)
        @inbounds maxnondiag, row, column = offdiagmax(A)
        @inbounds jacobirotate!(A, R, row, column)
        iterations += 1
    end
    R, iterations
end

function offdiagmax(x)
    # Finds the maximum off diagonal element in the lower quadrant
    # @Input: x, the matrix to find
    max = -1.0
    maxrow = 0
    maxcol = 0
    rows, columns = size(x)
    for column in 1:columns
        for row in column+1:rows
            row == column && continue
            y = abs(x[row, column])
            if y ≥ max
                max = y
                maxrow, maxcol = row, column
            end
        end
    end
    max, maxrow, maxcol
end


function jacobirotate!(A::Array{Float64}, R::Array{Float64}, k, l)
    s, c = 0.0, 0.0
    if A[k, l] ≠ 0.0
        τ = (A[l, l] - A[k, k])/(2A[k, l])
        if τ ≥ 0
            t = 1.0/(τ + √(1 + τ^2))
        else
            t = -1.0/(-τ + √(1 + τ^2))
        end
        c = 1.0/√(1 + t^2)
        s = c⋅t
    else
        c = 1.0
        s = 0.0
    end

    aₖₖ = A[k, k]
    aₗₗ = A[l, l]
    A[k, k] = c^2⋅aₖₖ - 2.0c⋅s⋅A[k, l] + s^2⋅aₗₗ
    A[l, l] = s^2⋅aₖₖ + 2.0c⋅s⋅A[k, l] + c^2⋅aₗₗ
    A[k, l], A[l, k] = 0.0, 0.0
    for i in 1:size(A, 1)
        if k ≠ i ≠ l
            aᵢₖ = A[i, k]
            aᵢₗ = A[i, l]
            A[i, k] = c⋅aᵢₖ - s⋅aᵢₗ
            A[k, i] = A[i, k]
            A[i, l] = c⋅aᵢₗ + s⋅aᵢₖ
            A[l, i] = A[i, l]
        end
        rᵢₖ = R[i, k]
        rᵢₗ = R[i, l]
        R[i, k] = c⋅rᵢₖ - s⋅rᵢₗ
        R[i, l] = c⋅rᵢₗ + s⋅rᵢₖ
    end
end



end

File: Project2/cpp/src/jacobi.cpp
#include <armadillo>
#include <cmath>
#include <ctime>
#include <chrono>
#include "jacobi.h"

void jacobi(arma::vec& eigval, arma::mat& eigvec, arma::mat& A){
  unsigned int n = A.n_cols;
  double maxiter = 1e8;               // Maximum number of iterations
  double epsilon = 1.0e-10;           // Tolerance for off-diagonal elements
  unsigned int iter = 0;
  unsigned int k, l;
  double max_offdiag = 1000.0;        // Set this larger than epsilon

  arma::mat A_copy = arma::mat(A);     // Make copy so input matrix is unchanged
  eigvec = arma::eye<arma::mat>(n, n); // Matrix that will hold eigenvalues

  auto start    = std::chrono::system_clock::now(); // Wall time
  std::clock_t startCPU = std::clock();                          // CPU clock start

  while(fabs(max_offdiag) > epsilon && static_cast<double>(iter) < maxiter){
    find_max_offdiag(A_copy, k, l, n);       // Find max offdiag element
    max_offdiag = A_copy(l,k);               // Fetch for next iteration
    jacobirotate(A_copy, eigvec, k, l, n);   // Perform rotation on largest element
    iter++;
  }
  // Output timing info
  auto end = std::chrono::system_clock::now();
  std::clock_t endCPU = std::clock();
  double elapsed_seconds_CPU = static_cast<double>(endCPU - startCPU)/CLOCKS_PER_SEC;
  std::chrono::duration<double> elapsed_seconds = end-start;
  std::cout << "Completed Jacobi eigenvalue algorithm with " << iter << " iterations" << std::endl;
  std::cout << "Elapsed wall time: " << elapsed_seconds.count() << "s" << std::endl;
  std::cout << "Elapsed CPU time: " << elapsed_seconds_CPU << "s" << std::endl;

  eigval = arma::sort(diagvec(A_copy));  // Diagonalized A holds eigenvalues
}

void jacobirotate(arma::mat& A, arma::mat& R, unsigned int& k, unsigned int& l, unsigned int& n){
  // Rotate: Find values of cos and sin
  double s,c;
  if(A(k,l) != 0.0){
    double t,tau;
    tau = (A(l,l) - A(k,k))/(2*A(k,l));
    if(tau>0){
      t = 1.0/(tau + sqrt(1.0 + tau*tau));
    }
    else{
      t = -1.0/(-tau + sqrt(1.0 + tau*tau));
    }

    c = 1/sqrt(1+t*t);
    s = c*t;
  }
  else{
    c = 1.0;
    s = 0.0;
  }
  double a_kk, a_ll, a_ik, a_il, r_ik, r_il;
  a_kk = A(k,k);
  a_ll = A(l,l);
  // Changing the matrix elements with indices k and l
  A(k,k) = c*c*a_kk - 2.0*c*s*A(k,l) + s*s*a_ll;
  A(l,l) = s*s*a_kk + 2.0*c*s*A(k,l) + c*c*a_ll;
  A(k,l) = 0.0;
  A(l,k) = 0.0;
  // Then we change the remaining elements
  for(unsigned int i = 0; i<n; i++){
    if( i != k && i != l){
      a_ik = A(i,k);
      a_il = A(i,l);
      A(i,k) = c*a_ik - s*a_il;
      A(k,i) = A(i,k);
      A(i,l) = c*a_il + s*a_ik;
      A(l,i) = A(i,l);
    }
  r_ik = R(i,k);
  r_il = R(i,l);
  R(i,k) = c*r_ik - s*r_il;
  R(i,l) = c*r_il + s*r_ik;
  }
}

void find_max_offdiag(arma::mat& A, unsigned int& k, unsigned int& l, unsigned int& n){
  // Finds maximum element on upper diagonal of a matrix A
  double max = 0.0;
  for(unsigned int i = 0; i < n; i++){
    for(unsigned int j = i+1; j < n; j++){
      if (fabs(A(i,j)) > max){
        max = fabs(A(i,j));
        l = i;
        k = j;
      }
    }
  }
}
